src/host/_stream.cpp

```diff
@@ -126,565 +165,286 @@ void AdjustCursorPosition(SCREEN_INFORMATION& screenInfo, _In_ til::point coordC
                                         _Inout_opt_ til::CoordType* const psScrollY)
+try
 {
-    const auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+    static constexpr wchar_t tabSpaces[8]{ L' ', L' ', L' ', L' ', L' ', L' ', L' ', L' ' };
+
     auto& textBuffer = screenInfo.GetTextBuffer();
     auto& cursor = textBuffer.GetCursor();
-    auto CursorPosition = cursor.GetPosition();
-    auto Status = STATUS_SUCCESS;
-    til::CoordType XPosition;
-    size_t TempNumSpaces = 0;
-    const auto fUnprocessed = WI_IsFlagClear(screenInfo.OutputMode, ENABLE_PROCESSED_OUTPUT);
-    const auto fWrapAtEOL = WI_IsFlagSet(screenInfo.OutputMode, ENABLE_WRAP_AT_EOL_OUTPUT);
-
-    // Must not adjust cursor here. It has to stay on for many write scenarios. Consumers should call for the
-    // cursor to be turned off if they want that.
-
-    const auto Attributes = screenInfo.GetAttributes();
-    const auto BufferSize = *pcb;
-    *pcb = 0;
-
-    auto lpString = pwchRealUnicode;
-
-    const auto coordScreenBufferSize = screenInfo.GetBufferSize().Dimensions();
-
-    static constexpr til::CoordType LOCAL_BUFFER_SIZE = 1024;
-    WCHAR LocalBuffer[LOCAL_BUFFER_SIZE];
+    const auto keepCursorVisible = WI_IsFlagSet(dwFlags, WC_KEEP_CURSOR_VISIBLE);
+    const auto wrapAtEOL = WI_IsFlagSet(screenInfo.OutputMode, ENABLE_WRAP_AT_EOL_OUTPUT);
+    auto it = pwchRealUnicode;
+    const auto end = it + *pcb / sizeof(wchar_t);
+    size_t numSpaces = 0;
 
-    while (*pcb < BufferSize)
+    if (cursor.IsDelayedEOLWrap() && wrapAtEOL)
     {
-        // correct for delayed EOL
-        if (cursor.IsDelayedEOLWrap() && fWrapAtEOL)
+        auto pos = cursor.GetPosition();
+        const auto delayed = cursor.GetDelayedAtPosition();
+        cursor.ResetDelayEOLWrap();
+        if (delayed == pos)
         {
-            const auto coordDelayedAt = cursor.GetDelayedAtPosition();
-            cursor.ResetDelayEOLWrap();
-            // Only act on a delayed EOL if we didn't move the cursor to a different position from where the EOL was marked.
-            if (coordDelayedAt.x == CursorPosition.x && coordDelayedAt.y == CursorPosition.y)
-            {
-                CursorPosition.x = 0;
-                CursorPosition.y++;
+            pos.x = 0;
+            pos.y++;
+            AdjustCursorPosition(screenInfo, pos, keepCursorVisible, psScrollY);
+        }
+    }
 
-                AdjustCursorPosition(screenInfo, CursorPosition, WI_IsFlagSet(dwFlags, WC_KEEP_CURSOR_VISIBLE), psScrollY);
+    if (WI_IsFlagClear(screenInfo.OutputMode, ENABLE_PROCESSED_OUTPUT))
+    {
+        numSpaces += _writeCharsLegacyUnprocessed(screenInfo, dwFlags, psScrollY, { it, end });
+        it = end;
+    }
 
-                CursorPosition = cursor.GetPosition();
-            }
+    while (it != end)
+    {
+        const auto nextControlChar = std::find_if(it, end, [](const auto& wch) { return !IS_GLYPH_CHAR(wch); });

```

Comment:
I bet you're going to vectorize this ;P

Reply:
Right now this function is a little faster than the VT parser, because it's less complex, but if that changes... 😄